WISH missing reflection due to tube gap

samueljackson92 opened the pull request:
**This must not be merged before #19207**

WISH users reported missing a reflection when using `PredictPeaks`. Current solution for now is to use a `tube-gap` parameter like CORELLI. 

**To test:**
 - The following script should fail before these changes and pass afterwards.

```python
import numpy as np

ws = LoadRaw(Filename='/archive/ndxwish/Instrument/data/cycle_16_5/WISH00038237.raw', OutputWorkspace='38237')
ws = ConvertUnits(ws, 'dSpacing', OutputWorkspace='38237')
UB = np.array([[-0.00601763,  0.07397297,  0.05865706],
                         [ 0.05373321,  0.050198,   -0.05651455],
                         [-0.07822144,  0.0295911,  -0.04489172]])
                         
SetUB(ws, UB=UB)

peaks = PredictPeaks(ws, WavelengthMin=0.1, WavelengthMax=100)
assert peaks.rowCount() == 510
filtered = FilterPeaks(peaks, "h^2+k^2+l^2", 75, '=')
assert filtered.rowCount() == 6
peak = filtered.row(2) 
assert  peak['h'] == -5 and peak['k'] == -1 and peak['l'] == -7
```

 - Check the new system test is passing (basically does the same as the above script)
 - Code review

Fixes #19315 .

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

AnthonyLim23 commented on Framework/API/test/DetectorSearcherTest.h:
```diff
@@ -0,0 +1,235 @@
+    for (size_t pointNo = 0; pointNo < info.size(); ++pointNo) {
+      const auto &det = info.detector(pointNo);
+      const auto q = convertDetectorPositionToQ(det);
+      checkResult(q, pointNo);
```
Am I right in thinking that this function just checks that the conversion from a point number to a  q value is correct?

samueljackson92 replied:
No. For each detector it converts the detector position to q, then searches for the detector using that q. We should get the same detector index back because that q should perfectly hit the detector.

AnthonyLim23 replied:
ok, thanks

AnthonyLim23 commented on Framework/Crystal/src/PredictPeaks.cpp:
```diff
@@ -470,40 +478,94 @@ void PredictPeaks::calculateQAndAddToOutput(const V3D &hkl,
+    // use extended detector space to try and guess peak position
+    const auto component =
+        m_inst->getComponentByName("extended-detector-space");
+    const auto c = boost::dynamic_pointer_cast<const ObjComponent>(component);
```
What is c? I would also avoid single character names as they are prone to typos.

samueljackson92 replied:
Yes

AnthonyLim23 commented on Framework/Crystal/test/PredictPeaksTest.h:
```diff
@@ -216,4 +247,54 @@ class PredictPeaksTest : public CxxTest::TestSuite {
 
+class PredictPeaksTestPerformance : public CxxTest::TestSuite {
+public:
+  void test_many_peaks_rectangular() {
```
shouldn't this be: test_manyPeaksRectangular

AnthonyLim23 replied:
should this function include an assert somewhere?

samueljackson92 replied:
Naming is a typo yes. 

It's a performance test, so we're checking for speed and not correctness. I could add a basic assertion but it shouldn't have detailed checking like a unit test.

AnthonyLim23 commented on Framework/Crystal/test/PredictPeaksTest.h:
```diff
@@ -216,4 +247,54 @@ class PredictPeaksTest : public CxxTest::TestSuite {
+    alg.execute();
+  }
+
+  void test_many_peaks() {
```
same as above

samueljackson92 replied:
Answers as above

AnthonyLim23 commented on Testing/SystemTests/tests/analysis/ISIS_WISHSingleCrystalReduction.py:
```diff
@@ -0,0 +1,54 @@
+
```
should the file end in Test.py?

samueljackson92 replied:
Yes.

AnthonyLim23 commented on Testing/SystemTests/tests/analysis/PredictPeaksTest.py:
```diff
@@ -89,9 +89,9 @@ def runTest(self):
                              CalculateStructureFactors=True)
 
-        self.assertEquals(peaks.getNumberPeaks(), 540)
+        self.assertEquals(peaks.getNumberPeaks(), 546)
```
why 546? Is it worth defining a constant for this script and then if it changes again it will only need to be edited in one place.

samueljackson92 replied:
I agree. Should be changed, but it was like that when I got here :P

AnthonyLim23 replied:
I know that one

AnthonyLim23 commented:
where is the file for the test? I have just tried to run the script and it said that it could not find the file.

DavidFair commented:
@AnthonyLim23 
If testing internally the easiest thing to do it change the path to `WISH00038237.raw` so it only reads `WISH00038237` and set **Search Data Archive** under *File-Manage User directories* to true.

***Note: You will need a machine with at least 16GB RAM (32 if possible) for these datasets***

AnthonyLim23 commented:
The script passed with this branch and failed on 3.9.2.

samueljackson92 commented:
@AnthonyLim23 hopefully my comments have addressed most of your points. I believe they're mostly typos etc. So could you potentially create a new issue and assign it to me to fix when I am back?

AnthonyLim23 commented:
@samueljackson92 I agree that they are mostly typos. I will talk to @DavidFair about them and see if he has time to fix them. If not then I will create an issue to fix the typos.

AnthonyLim23 commented:
Thanks @DavidFair.

AnthonyLim23 commented:
Once the test have passed I am happy to :shipit:

samueljackson92 commented:
Thank you very much @DavidFair!
